### PR TITLE
Move retry logic to the QueryBuilder

### PR DESCRIPTION
**Pull request by artonge**
Follow-up of https://github.com/nextcloud/server/pull/44504

### CODE REVIEW

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 1]
```diff
@@ -184,6 +184,13 @@ public function execute();
 	 */
 	public function executeQuery(): IResult;
 
+	/**
+	 * @since 30.0.0
+	 *
+	 * @param int $retryLimit - Retry the query up to $retryLimit times in case of a RetryableException. Initially equal to 4.
```
Comment (nickvergessen) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:190]: I don't think we should have retries by default :-X

Comment (artonge) [reply lib/public/DB/QueryBuilder/IQueryBuilder.php:190]: Unsure about it either, but note that this is only for `RetryableException`, we don't try forever, and there is a log, so it won't go unnoticed.

Not sure what the argument against would be, that is why I set the default to 4. But I do not have any strong opinion, so I am fine either way.

**Review by come-nc — APPROVED** [review]

File: lib/public/DB/QueryBuilder/IQueryBuilder.php (hunk 1 above)
Comment (juliushaertl) [reply lib/public/DB/QueryBuilder/IQueryBuilder.php:190]: Retries only work if the whole transaction is retried, so I'm also more in favor of making this an opt-in to not cause unexpected behaviour there when developers don't think about this and just call execute.

Comment (artonge) [reply lib/public/DB/QueryBuilder/IQueryBuilder.php:190]: > Retries only work if the whole transaction is retried

So it does not really make sense to have that logic here, no? I think I start to understand the value of https://github.com/nextcloud/server/blob/050c6d53b38d6737a49c13d4e6a20a6062160503/lib/public/AppFramework/Db/TTransactional.php#L94 :sweat_smile:

**Review by icewind1991 — CHANGES_REQUESTED** [review]
I think retrying by default is a bad idea and can lead to unexpected behavior for developers

File: lib/public/DB/QueryBuilder/IQueryBuilder.php (hunk 1 above)
Comment (artonge) [reply lib/public/DB/QueryBuilder/IQueryBuilder.php]: Default is now set to 0

**Pull request closed without merging**